File: iq/25/bounds/find.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package bounds

const indexError = -1

var BoundsError = Bounds{indexError, indexError}

type Bounds struct {
	Left  int
	Right int
}

func Find(nn []int, n int) Bounds {
	left := find(nn, n, findLeft)
	if left == -1 {
		return BoundsError
	}
	right := find(nn, n, findRight)
	if right == -1 {
		return BoundsError
	}
	return Bounds{Left: left, Right: right}
}

func find(nn []int, n int, f func(nn []int, n int, left, mid, right int) int) int {
	for left, right := 0, len(nn); left < right; {
		mid := left + (right-left)/2
		switch x := nn[mid]; {
		case n < x:
			right = mid
		case n > x:
			left = mid + 1
		default:
			return f(nn, n, left, mid, right)
		}
	}
	return indexError
}

func findLeft(nn []int, n int, left, mid, right int) int {
	i := find(nn[left:mid], n, findLeft)
	if i == indexError {
		return mid
	}
	return left + i
}

func findRight(nn []int, n int, left, mid, right int) int {
	i := find(nn[mid+1:right], n, findRight)
	if i == indexError {
		return mid
	}
	return mid + 1 + i
}

File: iq/25/bounds/find.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//
// clang-format off-next-line
//go:build ignore

#include "iq/25/bounds/find.h"

#include <functional>
#include <vector>

namespace iq::bounds {
namespace {

using FindFunc = std::function<int(const std::vector<int>&, int, std::size_t,
                                   std::size_t, std::size_t)>;

int find(const std::vector<int>& nn, int n, std::size_t left, std::size_t right,
         FindFunc fn) {
  while (left < right) {
    const std::size_t idx = left + (right - left) / 2;
    if (n < nn[idx]) {
      right = idx;
    } else if (n > nn[idx]) {
      left = idx + 1;
    } else {
      return fn(nn, n, left, idx, right);
    }
  }
  return kIllegalIndex;
}

int findLeft(const std::vector<int>& nn, int n, std::size_t left,
             std::size_t pivot, [[maybe_unused]] std::size_t right) {
  if (const int idx = find(nn, n, left, pivot, findLeft);
      idx != kIllegalIndex) {
    return idx;
  }
  return pivot;
}

int findRight(const std::vector<int>& nn, int n,
              [[maybe_unused]] std::size_t left, std::size_t pivot,
              std::size_t right) {
  if (const int idx = find(nn, n, pivot + 1, right, findRight);
      idx != kIllegalIndex) {
    return idx;
  }
  return pivot;
}

}  // namespace

Bounds Find(const std::vector<int>& nn, int n) {
  const int left = find(nn, n, 0, nn.size(), findLeft);
  if (left == kIllegalIndex) {
    return kIllegalBounds;
  }
  const int right = find(nn, n, left, nn.size(), findRight);
  if (right == kIllegalIndex) {
    return kIllegalBounds;
  }
  return {left, right};
}

}  // namespace iq::bounds
